node/actors/bft/src/leader/tests.rs

```diff
@@ -1,39 +1,394 @@
-use crate::testonly;
-use rand::{rngs::StdRng, Rng, SeedableRng};
-use zksync_concurrency::ctx;
-use zksync_consensus_roles::validator;
-
-// TODO(bruno): This only tests a particular case, not the whole method.
-#[tokio::test]
-async fn replica_commit() {
-    let ctx = &ctx::test_root(&ctx::RealClock);
-    let rng = &mut StdRng::seed_from_u64(6516565651);
-
-    let keys: Vec<_> = (0..1).map(|_| rng.gen()).collect();
-    let (genesis, val_set) = testonly::make_genesis(&keys, validator::Payload(vec![]));
-    let (mut consensus, _) = testonly::make_consensus(ctx, &keys[0], &val_set, &genesis).await;
-
-    consensus.leader.view = validator::ViewNumber(3);
-    consensus.leader.phase = validator::Phase::Commit;
-
-    let test_replica_msg =
-        consensus
-            .inner
-            .secret_key
-            .sign_msg(validator::ConsensusMsg::ReplicaCommit(
-                validator::ReplicaCommit {
-                    protocol_version: validator::CURRENT_VERSION,
-                    view: consensus.leader.view,
-                    proposal: rng.gen(),
-                },
-            ));
-
-    match consensus.leader.process_replica_commit(
-        ctx,
-        &consensus.inner,
-        test_replica_msg.cast().unwrap(),
-    ) {
-        Err(super::replica_commit::Error::UnexpectedProposal) => {}
-        res => panic!("unexpected result {res:?}"),
-    }
+//! ## Tests coverage
+//!
+//! - [x] replica_prepare_sanity
+//! - [x] replica_prepare_sanity_yield_leader_prepare
+//! - [x] replica_prepare_old_view
+//! - [x] replica_prepare_during_commit
+//! - [x] replica_prepare_not_leader_in_view
+//! - [x] replica_prepare_already_exists
+//! - [x] replica_prepare_num_received_below_threshold
+//! - [x] replica_prepare_invalid_sig
+//! - [x] replica_prepare_invalid_commit_qc
+//! - [x] replica_prepare_high_qc_of_current_view
+//! - [x] replica_prepare_high_qc_of_future_view
+//! - [x] replica_prepare_non_validator_signer // FAILS
+//!
+//! - [x] replica_commit_sanity
+//! - [x] replica_commit_sanity_yield_leader_commit
+//! - [x] replica_commit_old
+//! - [x] replica_commit_not_leader_in_view
+//! - [x] replica_commit_already_exists
+//! - [x] replica_commit_num_received_below_threshold
+//! - [x] replica_commit_invalid_sig
+//! - [x] replica_commit_unexpected_proposal
+//! - [x] replica_commit_protocol_version_mismatch // FAILS
+//!
+
+use super::{
+    replica_commit::Error as ReplicaCommitError, replica_prepare::Error as ReplicaPrepareError,
+};
+use crate::testonly::ut_harness::UTHarness;
+use assert_matches::assert_matches;
+use rand::Rng;
+use zksync_consensus_roles::validator::{
+    self, CommitQC, ConsensusMsg, LeaderCommit, LeaderPrepare, Phase, ProtocolVersion,
+    ReplicaCommit, ViewNumber,
+};
+
+#[tokio::test]
+async fn replica_prepare_sanity() {
```

Comment:
Nit: Cannot this be removed in favor of the following test? Ditto for other "sanity" tests.

Reply:
The following includes it, but I still separated for getting a better regression signals. I guess it can be removed.

Reply:
Seems redundant, yes. No need to separate them.

Reply:
I ended up using them differently: the first is for testing with a significant validator set size, the latter is for asserting the yielded message fields.